File: clients/JavaScript/model/entity-properties.js
const BuildProperties = require('./build-properties');
const AttackProperties = require('./attack-properties');
const RepairProperties = require('./repair-properties');
class EntityProperties {
    constructor(size, buildScore, destroyScore, canMove, populationProvide, populationUse, maxHealth, cost, sightRange, resourcePerHealth, build, attack, repair) {
        this.size = size;
        this.buildScore = buildScore;
        this.destroyScore = destroyScore;
        this.canMove = canMove;
        this.populationProvide = populationProvide;
        this.populationUse = populationUse;
        this.maxHealth = maxHealth;
        this.cost = cost;
        this.sightRange = sightRange;
        this.resourcePerHealth = resourcePerHealth;
        this.build = build;
        this.attack = attack;
        this.repair = repair;
    }
    static async readFrom(stream) {
        let size;
        size = await stream.readInt();
        let buildScore;
        buildScore = await stream.readInt();
        let destroyScore;
        destroyScore = await stream.readInt();
        let canMove;
        canMove = await stream.readBool();
        let populationProvide;
        populationProvide = await stream.readInt();
        let populationUse;
        populationUse = await stream.readInt();
        let maxHealth;
        maxHealth = await stream.readInt();
        let cost;
        cost = await stream.readInt();
        let sightRange;
        sightRange = await stream.readInt();
        let resourcePerHealth;
        resourcePerHealth = await stream.readInt();
        let build;
        if (await stream.readBool()) {
            build = await BuildProperties.readFrom(stream);
        } else {
            build = null;
        }
        let attack;
        if (await stream.readBool()) {
            attack = await AttackProperties.readFrom(stream);
        } else {
            attack = null;
        }
        let repair;
        if (await stream.readBool()) {
            repair = await RepairProperties.readFrom(stream);
        } else {
            repair = null;
        }
        return new EntityProperties(size, buildScore, destroyScore, canMove, populationProvide, populationUse, maxHealth, cost, sightRange, resourcePerHealth, build, attack, repair);
    }
    async writeTo(stream) {
        let size = this.size;
        await stream.writeInt(size);
        let buildScore = this.buildScore;
        await stream.writeInt(buildScore);
        let destroyScore = this.destroyScore;
        await stream.writeInt(destroyScore);
        let canMove = this.canMove;
        await stream.writeBool(canMove);
        let populationProvide = this.populationProvide;
        await stream.writeInt(populationProvide);
        let populationUse = this.populationUse;
        await stream.writeInt(populationUse);
        let maxHealth = this.maxHealth;
        await stream.writeInt(maxHealth);
        let cost = this.cost;
        await stream.writeInt(cost);
        let sightRange = this.sightRange;
        await stream.writeInt(sightRange);
        let resourcePerHealth = this.resourcePerHealth;
        await stream.writeInt(resourcePerHealth);
        let build = this.build;
        if (build === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await build.writeTo(stream);
        }
        let attack = this.attack;
        if (attack === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await attack.writeTo(stream);
        }
        let repair = this.repair;
        if (repair === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await repair.writeTo(stream);
        }
    }
}
module.exports = EntityProperties

File: clients/Ruby/model/entity_properties.rb
require_relative 'build_properties'
require_relative 'attack_properties'
require_relative 'repair_properties'
class EntityProperties
    attr_accessor :size
    attr_accessor :build_score
    attr_accessor :destroy_score
    attr_accessor :can_move
    attr_accessor :population_provide
    attr_accessor :population_use
    attr_accessor :max_health
    attr_accessor :cost
    attr_accessor :sight_range
    attr_accessor :resource_per_health
    attr_accessor :build
    attr_accessor :attack
    attr_accessor :repair
    def initialize(size, build_score, destroy_score, can_move, population_provide, population_use, max_health, cost, sight_range, resource_per_health, build, attack, repair)
        @size = size
        @build_score = build_score
        @destroy_score = destroy_score
        @can_move = can_move
        @population_provide = population_provide
        @population_use = population_use
        @max_health = max_health
        @cost = cost
        @sight_range = sight_range
        @resource_per_health = resource_per_health
        @build = build
        @attack = attack
        @repair = repair
    end
    def self.read_from(stream)
        size = stream.read_int()
        build_score = stream.read_int()
        destroy_score = stream.read_int()
        can_move = stream.read_bool()
        population_provide = stream.read_int()
        population_use = stream.read_int()
        max_health = stream.read_int()
        cost = stream.read_int()
        sight_range = stream.read_int()
        resource_per_health = stream.read_int()
        if stream.read_bool()
            build = BuildProperties.read_from(stream)
        else
            build = nil
        end
        if stream.read_bool()
            attack = AttackProperties.read_from(stream)
        else
            attack = nil
        end
        if stream.read_bool()
            repair = RepairProperties.read_from(stream)
        else
            repair = nil
        end
        EntityProperties.new(size, build_score, destroy_score, can_move, population_provide, population_use, max_health, cost, sight_range, resource_per_health, build, attack, repair)
    end
    def write_to(stream)
        stream.write_int(@size)
        stream.write_int(@build_score)
        stream.write_int(@destroy_score)
        stream.write_bool(@can_move)
        stream.write_int(@population_provide)
        stream.write_int(@population_use)
        stream.write_int(@max_health)
        stream.write_int(@cost)
        stream.write_int(@sight_range)
        stream.write_int(@resource_per_health)
        if @build.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @build.write_to(stream)
        end
        if @attack.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @attack.write_to(stream)
        end
        if @repair.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @repair.write_to(stream)
        end
    end
end
